### PR TITLE
Optimized how pool provided products are checked

### PR DESCRIPTION
**Pull request by Ceiu**
- Added bulk provided and derived provided product lookups to
  PoolCurator
- CandlepinPoolManager now makes use of the bulk provided and
  derived provided product lookups during various iterative
  operations

### CODE REVIEW

**Comment by Ceiu** [comment]
Optimization change, so not much in the way of specific tests; just verify that hypervisor checkins and whatnot are quicker.

However, there were some rebasing shenanigans, so it may be worth pointing out any odd-looking changes as there's a chance they're not entirely intentional.

**Comment by barnabycourt** [comment]
test this please

**Comment by Ceiu** [comment]
@candlepin-pull-request-bot retest this please

**Comment by Ceiu** [comment]
Still got some cases to fix on this one.

**Comment by Ceiu** [comment]
Should be good now, assuming I didn't overlook any other edge cases.

**Comment by Ceiu** [comment]
@candlepin-pull-request-bot retest this please

**Comment by kahowell** [comment]
retest this please

File: server/src/test/java/org/candlepin/resource/ConsumerResourceCreationTest.java [hunk 1]
```diff
@@ -45,7 +45,7 @@
 import org.candlepin.model.Owner;
 import org.candlepin.model.OwnerCurator;
 import org.candlepin.model.PermissionBlueprint;
-import org.candlepin.model.ProductCurator;
+import org.candlepin.model.PoolCurator;
 import org.candlepin.model.Release;
```
Comment (wottop) [inline server/src/test/java/org/candlepin/resource/ConsumerResourceCreationTest.java]: Same note as above.

**Comment by wottop** [comment]
looks good. ack.

**Comment by wottop** [comment]
spec test fail. 

buildbot: restest this please

**Comment by wottop** [comment]
ok, now its good.
